src/hardware/virtualbox.cpp

```diff
@@ -0,0 +1,628 @@
+	}
+}
+
+// ***************************************************************************
+// PCI card
+// ***************************************************************************
+
+struct PCI_VirtualBoxDevice : public PCI_Device {
+	enum : uint16_t {
+		vendor = 0x80ee,
+		device = 0xcafe,
+	};
+
+	PCI_VirtualBoxDevice() : PCI_Device(vendor, device) {}
+
+	bool InitializeRegisters(uint8_t registers[256]) override;
+	Bits ParseReadRegister(uint8_t regnum) override;
+	bool OverrideReadRegister(uint8_t regnum, uint8_t* rval,
+	                          uint8_t* rval_mask) override;
+	Bits ParseWriteRegister(uint8_t regnum, uint8_t value) override;
+};
+
+bool PCI_VirtualBoxDevice::InitializeRegisters(uint8_t registers[256])
+{
+	registers[0x04] = 0x01; // command register
+	registers[0x05] = 0x00;
+	registers[0x06] = 0x00; // status register
+	registers[0x07] = 0x00;
+
+	registers[0x08] = 0x00; // card revision
+	registers[0x09] = 0x00; // programming interface
+	registers[0x0a] = 0x00; // subclass code
+	registers[0x0b] = 0x00; // class code
+	registers[0x0c] = 0x00; // cache line size
+	registers[0x0d] = 0x00; // latency timer
+	registers[0x0e] = 0x00; // header type (other)
+
+	registers[0x3c] = 0xff; // no IRQ
+
+	constexpr auto port_num = static_cast<uint16_t>(port_num_virtualbox);
+	// BAR 0
+	registers[0x10] = static_cast<uint8_t>((port_num & 0xfc) + 1);
+	registers[0x11] = static_cast<uint8_t>((port_num >> 8) & 0xff);
+	registers[0x12] = 0;
+	registers[0x13] = 0;
+
+	return true;
+}
+
+Bits PCI_VirtualBoxDevice::ParseReadRegister(uint8_t regnum)
+{
+	return regnum;
+}
+
+bool PCI_VirtualBoxDevice::OverrideReadRegister([[maybe_unused]] uint8_t regnum,
+                                                [[maybe_unused]] uint8_t* rval,
+                                                [[maybe_unused]] uint8_t* rval_mask)
+{
+	return false;
+}
+
+Bits PCI_VirtualBoxDevice::ParseWriteRegister([[maybe_unused]] uint8_t regnum,
+                                              [[maybe_unused]] uint8_t value)
+{
+	return -1;
+}
+
+// ***************************************************************************
+// Lifecycle
+// ***************************************************************************
+
+void VIRTUALBOX_NotifyBooting()
+{
+	client_disconnect();
+}
+
+void VIRTUALBOX_Destroy(Section*)
+{
+	if (is_interface_enabled) {
+		client_disconnect();
+		PCI_RemoveDevice(PCI_VirtualBoxDevice::vendor,
+		                 PCI_VirtualBoxDevice::device);
+		IO_FreeWriteHandler(port_num_virtualbox, io_width_t::dword);
+		is_interface_enabled = false;
+	}
+}
+
+void VIRTUALBOX_Init(Section* sec)
+{
+	has_feature_mouse = MOUSEVMM_IsSupported(MouseVmmProtocol::VirtualBox);
+	if (has_feature_mouse) {
+		state.mouse_features.SetInitialValue();
+	}
+
+	// TODO: implement more features:
+	// - shared directories for VBSF.EXE driver by Javis Pedro
+	// - possibly display for the OS/2 Museum work-in-progres drivers
```

Comment:
Also, there is a 3D Win9x driver at [JHRobotics\SoftGPU](https://github.com/JHRobotics/softgpu#feature-support) - best suited for VirtualBox, but also for VMware and QEMU.

Reply:
OK, added a link. Please keep in mind that (unless someone steps in, determined to get it working) this is something for a VERY distant future, as the whole team is busy with gazillions of other important/interesting things.